File: code/0010. Regular Expression Matching/solution1-1.cpp
class Solution {
private:
    unordered_map<string, unordered_map<string, bool>> hashmap;

public:
    bool isMatch(string s, string p) {
        if (p.empty()) {
            return s.empty();
        }

        if (hashmap.count(s) > 0 && hashmap[s].count(p) > 0) {
            return hashmap[s][p];
        }

        bool matchFirst = !s.empty() && (p[0] == '.' || p[0] == s[0]);
        bool result;

        if (p.length() > 1 && p[1] == '*') {
            result = (isMatch(s, p.substr(2)) || (matchFirst && isMatch(s.substr(1), p)));
        }
        else {
            result = (matchFirst && isMatch(s.substr(1), p.substr(1)));
        }

        hashmap[s][p] = result;
        return result;
    }
};

File: code/0010. Regular Expression Matching/solution1-1.java
class Solution {
    private Map<String, Map<String, Boolean>> hashmap = new HashMap<>();
    
    public boolean isMatch(String s, String p) {
        if (p.isEmpty()) {
            return s.isEmpty();
        }
        
        if (hashmap.containsKey(s) && hashmap.get(s).containsKey(p)) {
            return hashmap.get(s).get(p);
        }
        
        boolean matchFirst = !s.isEmpty() && (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.');
        boolean result;
        
        if (p.length() > 1 && p.charAt(1) == '*') {
            result = isMatch(s, p.substring(2)) || (matchFirst && (isMatch(s.substring(1), p)));
        }
        else {
            result = matchFirst && isMatch(s.substring(1), p.substring(1));
        }
        
        if (!hashmap.containsKey(s)) {
            hashmap.put(s, new HashMap<>());
        }
        hashmap.get(s).put(p, result);
        
        return result;
    }
}